fcrepo-kernel-impl/src/main/java/org/fcrepo/kernel/impl/services/DeleteResourceServiceImpl.java

```diff
@@ -0,0 +1,122 @@
+
+import javax.inject.Inject;
+import java.util.stream.Stream;
+
+
+import org.fcrepo.kernel.api.ContainmentIndex;
+import org.fcrepo.kernel.api.Transaction;
+import org.fcrepo.kernel.api.exception.PathNotFoundException;
+import org.fcrepo.kernel.api.exception.RepositoryRuntimeException;
+import org.fcrepo.kernel.api.models.Binary;
+import org.fcrepo.kernel.api.models.Container;
+import org.fcrepo.kernel.api.models.FedoraResource;
+import org.fcrepo.kernel.api.models.ResourceFactory;
+import org.fcrepo.kernel.api.operations.DeleteResourceOperationFactory;
+import org.fcrepo.kernel.api.operations.ResourceOperation;
+import org.fcrepo.kernel.api.services.DeleteResourceService;
+import org.fcrepo.persistence.api.PersistentStorageSession;
+import org.fcrepo.persistence.api.PersistentStorageSessionManager;
+import org.fcrepo.persistence.api.exceptions.PersistentStorageException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This class mediates delete operations between the kernel and persistent storage layers
+ *
+ * @author dbernstein
+ */
+public class DeleteResourceServiceImpl implements DeleteResourceService {
+
+    private Logger LOGGER = LoggerFactory.getLogger(DeleteResourceService.class);
+
+    @Inject
+    private ContainmentIndex containmentIndex;
+
+    @Inject
+    private ResourceFactory resourceFactory;
+
+    @Inject
+    private DeleteResourceOperationFactory deleteResourceFactory;
+
+    @Inject
+    private PersistentStorageSessionManager psManager;
+
+    @Override
+    public void perform(final Transaction tx, final FedoraResource fedoraResource) {
+
+        final String fedoraResourceId = fedoraResource.getId();
+
+        try {
+            LOGGER.debug(format("deleting of %s", fedoraResourceId));
+            final PersistentStorageSession pSession = this.psManager.getSession(tx.getId());
+            deleteRecursivelyFromBottom(tx, pSession, fedoraResource);
+        } catch (final PersistentStorageException ex) {
+            throw new RepositoryRuntimeException(format("failed to delete resource %s", fedoraResourceId), ex);
+        }
+    }
+
+
+    private void deleteRecursivelyFromBottom(final Transaction tx, final PersistentStorageSession pSession,
+                                             final FedoraResource fedoraResource) throws PersistentStorageException {
+
+        final String fedoraId = fedoraResource.getId();
+
+        if (fedoraResource instanceof Container) {
+            final Stream<String> children = containmentIndex.getContainedBy(tx, fedoraResource);
+            children.forEach(childResourceId -> {
+                try {
+                    final FedoraResource res = resourceFactory.getResource(tx, childResourceId);
+                    deleteRecursivelyFromBottom(tx, pSession, res);
+                } catch (final PathNotFoundException ex) {
+                    LOGGER.warn(format("Path not found for %s: ignoring.", childResourceId));
```

Comment:
I'm trying to think through what it would mean for the path to not be found. The index is out of sync? Or a concurrent operation already deleted the item? Trying to figure out the severity of this kind of failure.

Also, use the logger's formatting syntax, so `"Path not found for {}: ignoring.", childResourceId`

Reply:
Yes - not sure myself.  We could throw a runtime exception here.